keps/sig-cli/3104-introduce-kuberc/README.md

```diff
@@ -0,0 +1,849 @@
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We propose introducing a new file that will be versioned and intended for user provided preferences. This new file is entirely opt-in.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+| Risk | Impact | Mitigation |
+| --- | --- | --- |
+| Confusing users with a new config file | Low | Documentation and education |
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The file will default to being located in `~/.kube/kuberc`. A flag will allow overriding this default location with a path i.e. `kubectl --kuberc /var/kube/rc`.
+
+Three initial top level keys are proposed.
+
+* `apiVersion` to determine the version of the config.
+* `commandAliases` for users to declare their own aliases for commands with flags and values.
+* `commandOverrides` for users to set default flags to apply to commands.
+
+`commandAliases` will take precedence when looking through the code execution path i.e. aliases -> builtin -> plugins. This will be expanded early in the kubectl invocation and used to replace argv. Additional flags and values will be appended to the end of the aliased command.
+
+`commandOverrides` is modeled after all configurable behavior being implemented as flags first. This is a design decision that was made after modeling out the intended behavior and realizing that targeting flags filled the use cases. A merge will be done in the execution of the command for flags with inline overrides taking precedence.
+
+```
+apiVersion: v1alpha1
+
+commandAliases:
+  - alias: getdbprod
+    command: get pods -l what=database --namespace us-2-production
```

Comment:
How do you see this command being executed? Is it just going to run whatever text you put here as a shell command?

Would there be a benefit to allowing variables to be substituted into the command, maybe just positionally like with a bash script ($1, $2, etc) or possibly named?  That is one of the limitations of bash aliases.

Should I be able to combine commands?  (ie. `kubectl wait --for=condition=Ready pod/foo && kubectl logs foo`)

I'm just throwing some questions out that come to mind when I look at this.  Maybe some of this is out of scope or at least out of scope initially.

Reply:
> How do you see this command being executed? Is it just going to run whatever text you put here as a shell command?

To be determined. I hope that doesn't wind up being the only option. I am thinking we might be able to get clever with Cobra exposing more.

> Would there be a benefit to allowing variables to be substituted into the command, maybe just positionally like with a bash script ($1, $2, etc) or possibly named? That is one of the limitations of bash aliases.

I like that idea. We can play with it when we go to implement.

> Should I be able to combine commands? (ie. kubectl wait --for=condition=Ready pod/foo && kubectl logs foo)

My initial thought is no, that's what bash aliases are for. You could bash alias multiple kubectl aliases :sweat_smile:.

Reply:
I'd again advise to start simple, w/o any replacements, and slowly iterate the bits we add.

Reply:
Added to open questions.